### PR TITLE
Add instructions for preparing a registry to use custom domains

### PR DESCRIPTION
**Pull request by avtakkar**

### CODE REVIEW

**Comment by sajayantony** [comment]
cc @SteveLas

**Comment by sajayantony** [comment]
@dlepow - can you help Review this? This is a on-boarding by request feature for **custom domain** support for ACR.

File: docs/custom-domain/registry-setup.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+## Prepare your Azure Key Vault
+For each domain, its TLS certificate must be added to an Azure Key Vault that is accessible by your registry. We recommend creating a new key vault containing only your TLS certificates and granting the registry's identity access to `get` secret.
+1. [Create](https://docs.microsoft.com/en-us/azure/key-vault/) a new Azure Key Vault.
+2. [Add](https://docs.microsoft.com/en-us/azure/key-vault/certificate-scenarios) your certificates to the key vault.
+3. Add an access policy to the key vault that grants your registry's identity access to `get` secret:\
+   `az keyvault set-policy --name <your-kv-name> --secret-permissions get --spn <registry-system-or-user-mi-client-id>`
+   - The output of the command to enable preview features on the registry will contain these client ids.
+
+For greater isolation, you may choose to put each certificate in its own key vault and set its access policy independently. The registry should have always have access to the TLS certificates.
```
Comment (sajayantony) [inline docs/custom-domain/registry-setup.md:64]: you may choose --> it is recommended
should have always have --> repeating words.

File: docs/custom-domain/registry-setup.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ ACR supports both user assigned and system assigned MI. If you choose to enable user assigned MI for your registry, create one first before proceeding. Instructions [here](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal).
+
+### Enable preview features
+1. `az login`
+2. `az account set -s <subscription-id-or-name> `
+3. `apiVersion="api-version=2019-12-01-preview"`
+4. `mgmtEndpoint=$(az cloud show --query endpoints.resourceManager -o tsv)`
+5. `accessToken=$(az account get-access-token --query accessToken -o tsv)`
+6. `registryResourceId=$(az acr show -n myregistry --query id -o tsv)`
+7. You can either enable a system assigned MI, or a user assigned MI, or both for your registry. Do one of the following:
```
Comment (sajayantony) [inline docs/custom-domain/registry-setup.md:43]: I think we should recommend system assigned identity incase KV is behind a VNET.

File: docs/custom-domain/registry-setup.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+## Prerequisites
+- [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/?view=azure-cli-latest), [curl](https://curl.haxx.se/) (or a similar tool)
+  - Consider using [Azure Cloud Shell](https://docs.microsoft.com/en-us/azure/cloud-shell/overview)
+- An Azure Container Registry. See [here](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-azure-cli) for instructions on how to create one.
+- Your custom domain names. The following two are required:
+  - custom registry domain to access the registry REST API: `container-registry.contoso.com` 
+  - custom data domain to access the registry content: `eastus-registry-data.contoso.com`
+    - Note that the custom data domain is region specific. For geo-replicated registries, each region should have its own custom data endpoint.
+
+  For each domain, you must prepare a single PEM formatted file containing the TLS private key and public certificate:
```
Comment (sajayantony) [inline docs/custom-domain/registry-setup.md:16]: Do you have openssl commands to create the certificate?

File: docs/custom-domain/registry-setup.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# Using Custom Domains with Azure Container Registry
+
+Every ACR is accessed using its login server. If you have a registry called `myregistry`, you access it using its default hostname, `myregistry.azurecr.io` (in Azure Public Cloud.) As a customer belonging to an organization, you may prefer to access your registry using a custom domain that is associated with your organization, for instance, `container-registry.contoso.com`.
+
+The following steps describe how you can achieve this.
+
+## Prerequisites
+- [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/?view=azure-cli-latest), [curl](https://curl.haxx.se/) (or a similar tool)
+  - Consider using [Azure Cloud Shell](https://docs.microsoft.com/en-us/azure/cloud-shell/overview)
+- An Azure Container Registry. See [here](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-azure-cli) for instructions on how to create one.
```
Comment (sajayantony) [inline docs/custom-domain/registry-setup.md:10]: Call out that the ACR should be a Premium registry

File: docs/custom-domain/registry-setup.md (hunk 1 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Updated

File: docs/custom-domain/registry-setup.md (hunk 2 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Updated

File: docs/custom-domain/registry-setup.md (hunk 3 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md:16]: Added an example

File: docs/custom-domain/registry-setup.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+8. `payload="{\"identity\": $identity, \"properties\": {\"dataEndpointEnabled\": true}}"`
+9. `endpoint="$mgmtEndpoint$registryResourceId?$apiVersion"`
+10. `curl -f -X PATCH -H "Authorization: Bearer $accessToken" -H "Content-Type: application/json" $endpoint -d "$payload"`
+
+## Prepare your Azure Key Vault
+For each domain, its TLS private key and public certificate pair must be added to an Azure Key Vault that is accessible by your registry as a single PEM formatted file. We recommend creating a new key vault containing only your TLS certificates and granting the registry's identity access to `get` secret.
+1. [Create](https://docs.microsoft.com/azure/key-vault/) a new Azure Key Vault.
+2. [Add](https://docs.microsoft.com/azure/key-vault/certificate-scenarios) your certificates to the key vault.
+3. Add an access policy to the key vault that grants your registry's identity access to `get` secret:\
+   `az keyvault set-policy --name <your-kv-name> --secret-permissions get --spn <registry-system-or-user-mi-client-id>`
```
Comment (dlepow) [inline docs/custom-domain/registry-setup.md:70]: Do you provide a command to obtain the id of the system-assigned managed identity?

File: docs/custom-domain/registry-setup.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ ACR supports both user assigned and system assigned MI. If you choose to enable user assigned MI for your registry, create one first before proceeding. Instructions [here](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal).
+
+### Enable preview features
+1. `az login`
+2. `az account set -s <subscription-id-or-name> `
+3. `apiVersion="api-version=2019-12-01-preview"`
+4. `mgmtEndpoint=$(az cloud show --query endpoints.resourceManager -o tsv)`
+5. `accessToken=$(az account get-access-token --query accessToken -o tsv)`
+6. `registryResourceId=$(az acr show -n myregistry --query id -o tsv)`
+7. You can either enable a system assigned MI, or a user assigned MI, or both for your registry. We recommend using a system assigned MI to enable advanced scenarios with virtual networks that are [called out below](#enhanced-security-with-virtual-networks). Do one of the following:
```
Comment (dlepow) [inline docs/custom-domain/registry-setup.md:52]: Do you want to call out the advanced scenario since you state it's not supported at the moment?

File: docs/custom-domain/registry-setup.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  -----END CERTIFICATE-----
+  ```
+  
+  For example, using [openssl](https://github.com/openssl/openssl):
+  - Create a self-signed public cert and private key
+     - `openssl req -nodes -x509 -newkey rsa:4096 -keyout container-registry.contoso.com.key.pem -out container-registry.contoso.com.cert.pem -days 365 -subj '/CN=container-registry.contoso.com/O=Contoso./C=US'`
+  - Create a single file containing both the public certificate and private key
+     - `cat container-registry.contoso.com.key.pem >> container-registry-contoso-com-pem`
+     - `cat container-registry.contoso.com.cert.pem >> container-registry-contoso-com-pem`
+
```
Comment (dlepow) [inline docs/custom-domain/registry-setup.md:33]: MIght want to say explicitly "Similarly, prepare a single PEM formatted file containing both the pubilc certificate and private key for each data domain."

File: docs/custom-domain/registry-setup.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  ```
+  
+  For example, using [openssl](https://github.com/openssl/openssl):
+  - Create a self-signed public cert and private key
+     - `openssl req -nodes -x509 -newkey rsa:4096 -keyout container-registry.contoso.com.key.pem -out container-registry.contoso.com.cert.pem -days 365 -subj '/CN=container-registry.contoso.com/O=Contoso./C=US'`
+  - Create a single file containing both the public certificate and private key
+     - `cat container-registry.contoso.com.key.pem >> container-registry-contoso-com-pem`
+     - `cat container-registry.contoso.com.cert.pem >> container-registry-contoso-com-pem`
+  
+  Azure Key Vault allows you to [create certificates](https://docs.microsoft.com/azure/key-vault/certificate-scenarios) signed with a Certificate Authority.
```
Comment (bariscaglar) [inline docs/custom-domain/registry-setup.md:34]: signed **by**?

File: docs/custom-domain/registry-setup.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  
+## Prepare your existing registry
+We will enable two features on your registry that are currently in preview:
+- Data Endpoints:\
+  This feature provides a dedicated endpoint for downloading content from your registry. If you have a registry in East US, on enabling this feature, a data endpoint is automatically created for you: `eastus.data.myregistry.azurecr.io`
```
Comment (bariscaglar) [inline docs/custom-domain/registry-setup.md:39]: This is not consistent with what is said above, with the dashes: eastus-data-myregistry.azurecr.io.

Comment (bariscaglar) [reply docs/custom-domain/registry-setup.md:39]: It is not necessarily wrong, but should we recommend the customers to use a similar approach as we do?

Comment (bariscaglar) [reply docs/custom-domain/registry-setup.md:39]: Also, if I have geo replication, are other data endpoints automatically created for me?

File: docs/custom-domain/registry-setup.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  
+- ACR Managed Identities:\
+  Managed Identities (MI) provide a mechanism to associate an Azure Active Directory identity with your registry, while relieving you of the burden of managing credentials. To learn more, see the documentation [here](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview).\
+ ACR supports both user assigned and system assigned MI. If you choose to enable user assigned MI for your registry, create one first before proceeding. Instructions [here](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal).
```
Comment (bariscaglar) [inline docs/custom-domain/registry-setup.md:43]: Will ACR create the system assigned for me? Calling that out makes it clear. "If you choose to enable system assigned... ACR will...."

File: docs/custom-domain/registry-setup.md (hunk 9 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md:39]: Which line above this are you referring to?

File: docs/custom-domain/registry-setup.md (hunk 10 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md:43]: You're right, this is a bit confusing. I'll update it.

File: docs/custom-domain/registry-setup.md (hunk 6 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: I've rephrased this a little. Does this make sense?

File: docs/custom-domain/registry-setup.md (hunk 8 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Rephrased this.

File: docs/custom-domain/registry-setup.md (hunk 10 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Does this make more sense?

File: docs/custom-domain/registry-setup.md (hunk 5 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Okay, added some commands

File: docs/custom-domain/registry-setup.md (hunk 9 above)
Comment (bariscaglar) [reply docs/custom-domain/registry-setup.md:40]: Line 13

Comment (avtakkar) [reply docs/custom-domain/registry-setup.md:40]: Line 13 provides an example for a custom data domain. The custom data domain could be anything, as chosen by a user.
This example is for a data endpoint that is ACR owned and automatically created for a registry when data endpoint is enabled.

Comment (avtakkar) [reply docs/custom-domain/registry-setup.md:40]: Also, for a custom owned domain called `contoso.com`, if more dots are prefixed to the name, that will also change the domain.

File: docs/custom-domain/registry-setup.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+     - Do _one_ of the following:
+       - To enable _only_ user assigned managed identity:
+         - `userMIResourceId=$(az identity show -n <user-mi-name> -g <user-mi-resource-group> --subscription <user-mi-subscription-id-or-name> --query id -o tsv)`
+         - `identity="{\"type\": \"userAssigned\", \"userAssignedIdentities\": {\"${userMIResourceId}\":{}}}"`
+       - To enable _both_ user and system assigned managed identities:
+         - `userMIResourceId=$(az identity show -n <user-mi-name> -g <user-mi-resource-group> --subscription <user-mi-subscription-id-or-name> --query id -o tsv)`
+         - `identity="{\"type\": \"systemAssigned,userAssigned\", \"userAssignedIdentities\": {\"${userMIResourceId}\":{}}}"`
+8. `payload="{\"identity\": $identity, \"properties\": {\"dataEndpointEnabled\": true}}"`
+9. `endpoint="$mgmtEndpoint$registryResourceId?$apiVersion"`
+10. `curl -f -X PATCH -H "Authorization: Bearer $accessToken" -H "Content-Type: application/json" $endpoint -d "$payload"`
```
Comment (yugangw-msft) [inline docs/custom-domain/registry-setup.md:67]: please use "az acr identity assign" to replace the the "curl"

Comment (avtakkar) [reply docs/custom-domain/registry-setup.md:67]: In addition to enabling identity, it is also enable data endpoints.

Comment (yugangw-msft) [reply docs/custom-domain/registry-setup.md:67]: I see. In the latest CLI, you can use "az acr update --data-endpoint-enabled" to enable data endpoint.

Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Thanks, updated!

File: docs/custom-domain/registry-setup.md (hunk 10 above)
Comment (bariscaglar) [reply docs/custom-domain/registry-setup.md]: Yes!

File: docs/custom-domain/registry-setup.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+     - `cat container-registry.contoso.com.key.pem >> container-registry-contoso-com-pem`
+     - `cat container-registry.contoso.com.cert.pem >> container-registry-contoso-com-pem`
+  - For each data domain, follow the same steps above to prepare the PEM formatted files containing the public certificate and private key.
+  
+  Azure Key Vault allows you to [create](https://docs.microsoft.com/azure/key-vault/certificate-scenarios) Certificate Authority (CA) signed certificates.
+  
+## Prepare your existing registry
+We will enable two features on your registry that are currently in preview:
+- Data Endpoints:\
+  This feature provides a dedicated endpoint for downloading content from your registry. If you have a registry in East US, on enabling this feature, a data endpoint is automatically created for you: `eastus.data.myregistry.azurecr.io`
```
Comment (bariscaglar) [inline docs/custom-domain/registry-setup.md:40]: I created data endpoints for a registry and this is what I got:

  "dataEndpointHostNames": [
    "[registryname].[region].data.azurecr.io"
  ]

This suggests it would be [region].data.[registryname].azurecr.io

Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Addressed

File: docs/custom-domain/registry-setup.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  For example, using [openssl](https://github.com/openssl/openssl):
+  - Create a self-signed public cert and private key
+     - `openssl req -nodes -x509 -newkey rsa:4096 -keyout container-registry.contoso.com.key.pem -out container-registry.contoso.com.cert.pem -days 365 -subj '/CN=container-registry.contoso.com/O=Contoso./C=US'`
+  - Create a single file containing both the public certificate and private key
+     - `cat container-registry.contoso.com.key.pem >> container-registry-contoso-com-pem`
+     - `cat container-registry.contoso.com.cert.pem >> container-registry-contoso-com-pem`
+  - For each data domain, follow the same steps above to prepare the PEM formatted files containing the public certificate and private key.
+  
+  Azure Key Vault allows you to [create](https://docs.microsoft.com/azure/key-vault/certificate-scenarios) Certificate Authority (CA) signed certificates.
+
```
Comment (bariscaglar) [inline docs/custom-domain/registry-setup.md:36]: When I followed the document, I created a PKCS12 since that's the AKV default. This pitfall can be called out here. I know that you already mention PEM, but using Azure, they might be under the impression that what they create will work out of the box.

Comment (avtakkar) [reply docs/custom-domain/registry-setup.md]: Thanks, called this out.

File: docs/custom-domain/registry-setup.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  - Create a single file containing both the public certificate and private key
+     - `cat container-registry.contoso.com.key.pem >> container-registry-contoso-com-pem`
+     - `cat container-registry.contoso.com.cert.pem >> container-registry-contoso-com-pem`
+  - For each data domain, follow the same steps above to prepare the PEM formatted files containing the public certificate and private key.
+  
+  Azure Key Vault allows you to [create](https://docs.microsoft.com/azure/key-vault/certificate-scenarios) Certificate Authority (CA) signed certificates. 
+  - If you choose to use the Azure Portal to create the certificates, be sure to select certificate content type as PEM.
+ 
+## Prepare your existing registry
+We will enable two features on your registry that are currently in preview:
```
Comment (yugangw-msft) [inline docs/custom-domain/registry-setup.md:39]: Data Endpoint and Managed identities are GA now

File: docs/custom-domain/README.md [hunk 15]
```diff
@@ -1,56 +1,96 @@
...
+## Prepare your existing registry
+We will enable two features on your registry:
+- Data Endpoints:\
+  This feature provides a dedicated endpoint for downloading content from your registry. If you have a registry in East US, on enabling this feature, a data endpoint is automatically created for you: `myregistry.eastus.data.azurecr.io`
+  
+- ACR Managed Identities:\
+  Managed Identities provide a mechanism to associate an Azure Active Directory identity with your registry, while relieving you of the burden of managing credentials. To learn more, see the documentation [here](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview).\
+ ACR supports both user assigned and system assigned managed identities.
+
+### Enable preview features
```
Comment (dlepow) [inline docs/custom-domain/README.md:47]: Maybe reword heading: "Enable data endpoints and managed identities"

File: docs/custom-domain/README.md [hunk 16]
```diff
@@ -1,56 +1,96 @@
...
+2. `az account set -s <subscription-id-or-name> `
+3. `az acr update --data-endpoint-enabled true -n myregistry`
+4. You can either enable a system assigned managed identity, a user assigned managed identity, or both for your registry. We recommend using system assigned managed identity to enable advanced scenarios with virtual networks that, although not supported currently, are [coming soon](#enhanced-security-with-virtual-networks). Do _one_ of the following:
+   - To enable only system assigned managed identity: 
+     - `az acr identity assign -n myregistry --identities [system]`
+   - To enable user assigned managed identity, with or without a system identity: 
+     - Create a user assigned managed identity following the instructions [here](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal).
+     - Do _one_ of the following:
+       - To enable _only_ user assigned managed identity:
+         - `az acr identity assign -n avtakkarweu --identities "<arm-resource-id-of-user-assigned-identity>"`
```
Comment (dlepow) [inline docs/custom-domain/README.md:58]: Change registry name to myregistry here and in next step

File: docs/custom-domain/README.md [hunk 17]
```diff
@@ -1,56 +1,96 @@
...
+       - To enable _both_ user and system assigned managed identities:
+         - `az acr identity assign -n avtakkarweu --identities "<arm-resource-id-of-user-assigned-identity>" [system]`
+
+## Prepare your Azure Key Vault
+For each domain, its TLS private key and public certificate pair must be added to an Azure Key Vault that is accessible by your registry as a single PEM formatted file. We recommend creating a new key vault containing only your TLS certificates and granting the registry's identity access to `get` secret.
+1. [Create](https://docs.microsoft.com/azure/key-vault/) a new Azure Key Vault.
+2. [Add](https://docs.microsoft.com/azure/key-vault/certificate-scenarios) your certificates to the key vault.
+3. Add an access policy to the key vault that grants your registry's identity access to `get` secret:\
+   `az keyvault set-policy --name <your-kv-name> --secret-permissions get --spn <registry-system-or-user-mi-principal-id>`
+   - The output of the command to enable preview features on the registry will contain the principal ids of the assiged identities.
```
Comment (dlepow) [inline docs/custom-domain/README.md:68]: "...to enable identities on the registry..."?

File: docs/custom-domain/README.md [hunk 18]
```diff
@@ -1,56 +1,96 @@
...
+
+### Enhanced security with Virtual Networks
+Azure Key Vault allows you to [restrict access](https://docs.microsoft.com/azure/key-vault/key-vault-overview-vnet-service-endpoints) to specific virtual networks only. ACR custom domains are currently _not supported_ where key vault access is restricted, but this is work in progress and will be available with system managed identities only.
+   
+## Prepare your DNS zone
+1. The custom registry domain must have a CNAME record with the target registry login server:\
+   `container-registry.contoso.com` --> `myregistry.azurecr.io`
+2. The regional custom data domain must have a CNAME record with the target regional registry data endpoint:\
+   `eastus-registry-data.contoso.com` --> `myregistry.eastus.data.azurecr.io`
+   - The output of the command to enable preview features on the registry will contain the regional data endpoint.
```
Comment (dlepow) [inline docs/custom-domain/README.md:85]: ..."to enable data endpoints on the registry.."?

File: docs/custom-domain/registry-setup-deprecated.md [hunk 19]
```diff
@@ -0,0 +1,56 @@
+# How to use a custom domain for azure container registry
+
```
Comment (dlepow) [inline docs/custom-domain/registry-setup-deprecated.md:2]: Is the content in this article still needed?

File: docs/custom-domain/README.md [hunk 20]
```diff
@@ -1,56 +1,96 @@
...
+  -----BEGIN PRIVATE KEY-----  
+  .....  
+  -----END PRIVATE KEY-----  
+  -----BEGIN CERTIFICATE-----  
+  .....  
+  -----END CERTIFICATE-----
+  ```
+  
+  For example, using [openssl](https://github.com/openssl/openssl):
+  - Create a self-signed public cert and private key
```
Comment (sajayantony) [inline docs/custom-domain/README.md:28]: >private key [](start = 41, length = 11)

Just highlight these are bash command on the top somehwere. 
Also break up it up as a multi-line command with \ etc.

File: docs/custom-domain/registry-setup-deprecated.md (hunk 19 above)
Comment (avtakkar) [reply docs/custom-domain/registry-setup-deprecated.md:2]: @sajayantony ?

File: docs/custom-domain/README.md (hunk 20 above)
Comment (avtakkar) [reply docs/custom-domain/README.md:28]: Marked the code as `shell` and split up the lines.

File: docs/custom-domain/README.md (hunk 15 above)
Comment (avtakkar) [reply docs/custom-domain/README.md]: Done

File: docs/custom-domain/registry-setup-deprecated.md (hunk 19 above)
Comment (sajayantony) [reply docs/custom-domain/registry-setup-deprecated.md]: Yes. that should be removed since we now support custom domains through a support ticket. I would rather we remove the nginx doc delete the template/commands.

Comment (avtakkar) [reply docs/custom-domain/registry-setup-deprecated.md]: Okay, removed
